### DISCUSSION TITLE
Welcome to the Hyprspace Discussion Forum!

### DISCUSSION THREAD

**Discussion by alecbcs** [alecbcs/hyprspace#3] [Announcements]
## 👋 Hello and Welcome!
 Please feel free to:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by Winterhuman** [comment]
Does hyprspace use Wireguard as the underlying VPN once two peers find each other or is it a separate VPN implementation all together?

**Reply by alecbcs** [reply]
This is a great question! So Hyprspace doesn't use Wireguard. We use [LibP2P](https://github.com/libp2p) to setup the connections between peers but LibP2P uses an authentication/public-private key encryption system that is very near identical to what is used in Wireguard. I tested out using Wireguard for Hyprspace originally but found that the performance of the `wireguard-go` was less than using straight encrypted libp2p.

**Reply by Winterhuman** [reply]
Hyprspace performance was actually better than Wireguard? Wow, well I guess that addresses performance concerns compared to existing VPN solutions!

**Reply by alecbcs** [reply]
Wireguard built into the Linux Kernel is still definitely faster than Hyprspace. But for userspace implementations written in Go I've been really happy with Hyprspace pulling out almost 500Mbps between my laptop and development server which was much faster than wireguard-go in my testing.

**Comment by Winterhuman** [comment]
Another question I thought of while reading about [pcp](https://github.com/dennis-tra/pcp), does hyprspace only accept ipv4 when giving peers ip-addresses? Does the initial connection between two devices use ipv6 first with ipv4 as a fallback?

**Reply by alecbcs** [reply]
This is a good question! I believe that externally between peers their IPs can be either IPv4 or IPv6. In the case that both are present, I believe `libp2p` will actually publish that the node is available on both and it's up to the other clients which to choose/they support.

**Comment by Winterhuman** [comment]
Quick question, how can I use `hyprspace` to forward traffic from one peer to another? let's say the following is the case:

(Peer-A on Network A) and (Peer-B on Network B)

I connect the two peers via `hyprspace`, how would I forward the vpn connection so that Peer-A will have some or all of it's traffic forwarded to Peer-B/Network-B?

**Reply by alecbcs** [reply]
You can either use a couple of IP table rules to relay traffic from one interface to another, (I've done this in the past with something like [this](https://unix.stackexchange.com/questions/76300/iptables-port-to-another-ip-port-from-the-inside)), or you can use the [hyperspace relay](https://github.com/hyprspace/relay) which I built & use for a very similar purpose for relaying requests from a port on a public IP through hyperspace to a peer on a different network.

**Reply by mrbluecoat** [reply]
Can hyprspace work as a transparent proxy? e.g. instead of Peer-A traffic flowing _to_ Peer-B, it flows _through_ Peer-B and on to the Internet

**Reply by Winterhuman** [reply]
@mrbluecoat I believe that's what the above does, using the IP table rules the traffic from Peer-A would appear to come from Peer-B instead thus acting as a proxy

**Reply by mrbluecoat** [reply]
I have a home Peer-A on a private NAT'd network A that can access the internet through a home gateway router. I have a public VPS Peer-B on Network B. I can install Hyprspace and ping each other on 10.1.1.x. Neither can access the Internet on interface utun#. I installed TinyProxy on the public VPS and can proxy through it using its public IP address but not its Hyprspace 10.1.1.x address. I'm probably missing something obvious so apologies for the n00b question.

**Reply by alecbcs** [reply]
@mrbluecoat hahaha at some level we're all n00bs so no worries! Just so that I can understand your use case better are you wanting to forward requests from the VPS to your home Peer-A? Or you want to be able to forward your traffic from Peer-A out of the VPS like a traditional VPN service?

**Reply by mrbluecoat** [reply]
I figured I'd start with the more traditional use case of a public VPS as an exit node like a traditional VPN and then if that worked try the more unusual use case of externally tunneling into my home network and using it as an exit node (for edge cases like watching Netflix when traveling abroad)

**Reply by mrbluecoat** [reply]
For the traditional VPN scenario, here's the basic flow:
Client running hyprspace on a home network behind a NAT requests example.com > Server running hyprspace as a transparent proxy and peered with client receives the request and does a GET example.com on behalf of client > GET result sent back to client

**Reply by mrbluecoat** [reply]
Maybe I'm just confusing TUN and TAP and this is the wrong tool for this use case.

**Reply by mrbluecoat** [reply]
For anyone coming across this thread, it turns out `iptables` was only half the solution. [I needed a transparent forward proxy to complete my use case](https://serverfault.com/questions/1084177/route-all-internet-traffic-through-tun-interface-to-remote-proxy).
